cli/cmd/capture/download.go

```
@@ -52,8 +73,115 @@ var (
 	ErrFileNotAccessible         = errors.New("file does not exist or is not readable")
 	ErrEmptyDownloadOutput       = errors.New("download command produced no output")
 	ErrFailedToCreateDownloadPod = errors.New("failed to create download pod")
+	ErrUnsupportedNodeOS         = errors.New("unsupported node operating system")
 )
 
+// DownloadCmd holds all OS-specific commands and configurations
+type DownloadCmd struct {
+	ContainerImage   string
+	SrcFilePath      string
+	MountPath        string
+	KeepAliveCommand []string
+	FileCheckCommand []string
+	FileReadCommand  []string
+}
+
+// DownloadService encapsulates the download functionality and shared dependencies
+type DownloadService struct {
+	kubeClient kubernetes.Interface
+	config     *rest.Config
+	namespace  string
+	ctx        context.Context
+}
+
+// NewDownloadService creates a new download service with shared dependencies
+func NewDownloadService(ctx context.Context, kubeClient kubernetes.Interface, config *rest.Config, namespace string) *DownloadService {
+	return &DownloadService{
+		kubeClient: kubeClient,
+		config:     config,
+		namespace:  namespace,
+		ctx:        ctx,
+	}
+}
+
+func getDownloadCmd(node *corev1.Node, hostPath, fileName string) *DownloadCmd {
+	nodeOS, err := getNodeOS(node)
+	if err != nil {
+		retinacmd.Logger.Error("Failed to detect node OS", zap.String("node", node.Name), zap.Error(err))
+		return nil
+	}
```

isn't it better to return the error and leave to the caller to decide what to do with it? Or if you don't want an empty output to be considered an error, change getNodeOS to return only string/NodeOS as getDownloadCmd is doing and the caller will decide if that's acceptable or not. It does not look consistent to me as it is now because getNodeOS outputs an error -> getDownloadCmd does not consider that error and returns empty string -> DownloadFile creates an error again based on empty output.

---

I agree. I have updated this to return a nil and the error